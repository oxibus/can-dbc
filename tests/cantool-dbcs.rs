#![cfg(feature = "with-serde")]

use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use std::fs::File;
use std::io::Read as _;
use std::path::Path;
use std::{fs, io};

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory, using the `can_dbc` crate.
#[test]
fn test_cantools_dbc_files() {
    // Configure insta snapshot content and location
    with_settings! {
        { omit_expression => true,
          snapshot_path => "snapshots-cantools-dbcs",
          prepend_module_to_snapshot => false },
        {
            test_cantools_dbc_files_int().unwrap();
        }
    }
}

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory.
fn test_cantools_dbc_files_int() -> io::Result<()> {
    for dbc_path in fs::read_dir("./tests/cantools-dbcs")? {
        let path = dbc_path?.path();
        if path
            .extension()
            .and_then(|s| s.to_str())
            .map(|s| s.eq_ignore_ascii_case("dbc"))
            .is_some()
        {
            parse_one_file(&path)?;
        }
    }
    Ok(())
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) -> io::Result<()> {
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    eprintln!("Testing DBC file: {file_name}.dbc");
    let mut buffer = Vec::new();
    File::open(path)?.read_to_end(&mut buffer)?;

    match can_dbc::DBC::from_slice(&buffer) {
        Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
        Err(e) => {
            eprintln!("Failed to parse {file_name}.dbc: {e:?}");
            assert_debug_snapshot!(format!("!error___{file_name}"), e);
        }
    }
    Ok(())
}
