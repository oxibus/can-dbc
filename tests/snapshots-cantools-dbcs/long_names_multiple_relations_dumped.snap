---
source: tests/cantool-dbcs.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - - Sender_2_aaaaaaaaaaaaaaaaaaaaaaa
    - Receiver_2_zzzzzzzzzzzzzzzzzzzzz
    - Receiver_1
    - Node_6789_123456789_123456789_12
value_tables:
  - value_table_name: Value_Table_short
    value_descriptions:
      - a: 1
        b: "Very long, long, long description for the value '0x1'"
      - a: 0
        b: "Very long, long, long description for the value '0x0'"
  - value_table_name: Value_Table_cdefghi_ABCDEFGHI_AB
    value_descriptions:
      - a: 3
        b: "Description for the value '0x3'"
      - a: 2
        b: "Description for the value '0x2'"
      - a: 1
        b: "Description for the value '0x1'"
      - a: 0
        b: "Description for the value '0x0'"
messages:
  - message_id:
      Standard: 85
    message_name: Msg_Long_Name_56789_123456789_12
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: rx_twice_short
        multiplexer_indicator: Plain
        start_bit: 16
        signal_size: 18
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: rx_twice_11111111111111111111111
        multiplexer_indicator: Plain
        start_bit: 8
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Sig_used_twice_efgh_abcdefghi_ab
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 6
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 6
    message_name: TX_twice
    message_size: 2
    transmitter:
      NodeName: Node_6789_123456789_123456789_12
    signals:
      - name: rx_twice_long_yyyyyyyyyyyyyyyyyy
        multiplexer_indicator: Plain
        start_bit: 8
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Receiver_1
          - Receiver_2_zzzzzzzzzzzzzzzzzzzzz
      - name: rx_twice_short
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Receiver_2_zzzzzzzzzzzzzzzzzzzzz
          - Receiver_1
  - message_id:
      Standard: 5
    message_name: RX_TX_1
    message_size: 8
    transmitter:
      NodeName: Node_6789_123456789_123456789_12
    signals:
      - name: Sig_used_twice_efgh_abcdefg_0000
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 9
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 4
    message_name: MSG_CASE_TEST
    message_size: 8
    transmitter: VectorXXX
    signals: []
  - message_id:
      Standard: 3
    message_name: msg_case_test
    message_size: 8
    transmitter: VectorXXX
    signals: []
  - message_id:
      Standard: 2
    message_name: Msg_with_value_table_sigs
    message_size: 3
    transmitter: VectorXXX
    signals:
      - name: Sig_with_short_val_table
        multiplexer_indicator: Plain
        start_bit: 16
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Sig_with_long_val_table_2
        multiplexer_indicator: Plain
        start_bit: 8
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Sig_with_long_val_table_1
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 1
    message_name: Msg_Long_Name_56789_1234567_0000
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: rx_twice_11111111111111111111111
        multiplexer_indicator: Plain
        start_bit: 8
        signal_size: 2
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Sig_used_twice_efgh_abcdefg_0001
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 0
    message_name: Msg_Long_Name_56789_1234567_0001
    message_size: 8
    transmitter: VectorXXX
    signals:
      - name: rx_twice_short
        multiplexer_indicator: Plain
        start_bit: 8
        signal_size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Sig_used_twice_efgh_abcdefg_0002
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 6
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
message_transmitters:
  - message_id:
      Standard: 6
    transmitter:
      - NodeName: Node_6789_123456789_123456789_12
      - NodeName: Sender_2_aaaaaaaaaaaaaaaaaaaaaaa
environment_variables: []
environment_variable_data: []
signal_types: []
comments: []
attribute_definitions:
  - Signal: "\"GenSigSendType\" ENUM  \"Cyclic\",\"OnWrite\",\"OnWriteWithRepetition\",\"OnChange\",\"OnChangeWithRepetition\",\"IfActive\",\"IfActiveWithRepetition\",\"NoSigSendType\""
  - Signal: "\"GenSigInactiveValue\" INT 0 0"
  - Message: "\"GenMsgCycleTime\" INT 0 0"
  - Message: "\"GenMsgSendType\" ENUM  \"Cyclic\",\"not_used\",\"not_used\",\"not_used\",\"not_used\",\"Cyclic\",\"not_used\",\"IfActive\",\"NoMsgSendType\""
  - Node: "\"NmStationAddress\" HEX 0 0"
  - Plain: "\"DBName\" STRING "
  - Plain: "\"BusType\" STRING "
  - Signal: "\"SystemSignalLongSymbol\" STRING "
  - Message: "\"SystemMessageLongSymbol\" STRING "
  - Node: "\"SystemNodeLongSymbol\" STRING "
attribute_defaults:
  - attribute_name: GenSigSendType
    attribute_value:
      AttributeValueCharString: Cyclic
  - attribute_name: GenSigInactiveValue
    attribute_value:
      AttributeValueF64: 0
  - attribute_name: GenMsgCycleTime
    attribute_value:
      AttributeValueF64: 0
  - attribute_name: GenMsgSendType
    attribute_value:
      AttributeValueCharString: NoMsgSendType
  - attribute_name: NmStationAddress
    attribute_value:
      AttributeValueF64: 0
  - attribute_name: DBName
    attribute_value:
      AttributeValueCharString: ""
  - attribute_name: BusType
    attribute_value:
      AttributeValueCharString: CAN
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      AttributeValueCharString: ""
  - attribute_name: SystemMessageLongSymbol
    attribute_value:
      AttributeValueCharString: ""
  - attribute_name: SystemNodeLongSymbol
    attribute_value:
      AttributeValueCharString: ""
attribute_values:
  - attribute_name: DBName
    attribute_value:
      RawAttributeValue:
        AttributeValueCharString: long_names_multiple_relations
  - attribute_name: SystemNodeLongSymbol
    attribute_value:
      NetworkNodeAttributeValue:
        - Sender_2_aaaaaaaaaaaaaaaaaaaaaaa
        - AttributeValueCharString: Sender_2_aaaaaaaaaaaaaaaaaaaaaaaAAAAAA
  - attribute_name: SystemNodeLongSymbol
    attribute_value:
      NetworkNodeAttributeValue:
        - Receiver_2_zzzzzzzzzzzzzzzzzzzzz
        - AttributeValueCharString: Receiver_2_zzzzzzzzzzzzzzzzzzzzzZZZ
  - attribute_name: SystemNodeLongSymbol
    attribute_value:
      NetworkNodeAttributeValue:
        - Node_6789_123456789_123456789_12
        - AttributeValueCharString: Node_6789_123456789_123456789_123456789
  - attribute_name: SystemMessageLongSymbol
    attribute_value:
      MessageDefinitionAttributeValue:
        - Standard: 85
        - AttributeValueCharString: Msg_Long_Name_56789_123456789_123456789_Copy_2
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      SignalAttributeValue:
        - Standard: 85
        - Sig_used_twice_efgh_abcdefghi_ab
        - AttributeValueCharString: Sig_used_twice_efgh_abcdefghi_abcdefghi_abcdefghi
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      SignalAttributeValue:
        - Standard: 6
        - rx_twice_long_yyyyyyyyyyyyyyyyyy
        - AttributeValueCharString: rx_twice_long_yyyyyyyyyyyyyyyyyyYYY
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      SignalAttributeValue:
        - Standard: 5
        - Sig_used_twice_efgh_abcdefg_0000
        - AttributeValueCharString: Sig_used_twice_efgh_abcdefghi_abcdefghi_abcdefghi
  - attribute_name: SystemMessageLongSymbol
    attribute_value:
      MessageDefinitionAttributeValue:
        - Standard: 1
        - AttributeValueCharString: Msg_Long_Name_56789_123456789_123456789_Copy_1
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      SignalAttributeValue:
        - Standard: 1
        - Sig_used_twice_efgh_abcdefg_0001
        - AttributeValueCharString: Sig_used_twice_efgh_abcdefghi_abcdefghi_abcdefghi
  - attribute_name: SystemMessageLongSymbol
    attribute_value:
      MessageDefinitionAttributeValue:
        - Standard: 0
        - AttributeValueCharString: Msg_Long_Name_56789_123456789_123456789
  - attribute_name: SystemSignalLongSymbol
    attribute_value:
      SignalAttributeValue:
        - Standard: 0
        - Sig_used_twice_efgh_abcdefg_0002
        - AttributeValueCharString: Sig_used_twice_efgh_abcdefghi_abcdefghi_abcdefghi
value_descriptions:
  - Signal:
      message_id:
        Standard: 2
      signal_name: Sig_with_short_val_table
      value_descriptions:
        - a: 1
          b: "Very long, long, long description for the value '0x1'"
        - a: 0
          b: "Very long, long, long description for the value '0x0'"
  - Signal:
      message_id:
        Standard: 2
      signal_name: Sig_with_long_val_table_2
      value_descriptions:
        - a: 13
          b: "value '0xD'"
        - a: 12
          b: "Dvalue '0xC'"
        - a: 11
          b: "value '0xB'"
        - a: 10
          b: "value '0xA'"
        - a: 9
          b: "value '0x9'"
        - a: 8
          b: "value '0x8'"
        - a: 7
          b: "value '0x7'"
        - a: 6
          b: "value '0x6'"
        - a: 5
          b: "value '0x5'"
        - a: 4
          b: "value '0x4'"
        - a: 3
          b: "value '0x3'"
        - a: 2
          b: "value '0x2'"
        - a: 1
          b: "value '0x1'"
        - a: 0
          b: "value '0x0'"
  - Signal:
      message_id:
        Standard: 2
      signal_name: Sig_with_long_val_table_1
      value_descriptions:
        - a: 3
          b: "Description for the value '0x3'"
        - a: 2
          b: "Description for the value '0x2'"
        - a: 1
          b: "Description for the value '0x1'"
        - a: 0
          b: "Description for the value '0x0'"
signal_type_refs: []
signal_groups: []
signal_extended_value_type_list: []
extended_multiplex: []
